### DISCUSSION TITLE
append command does not add lines into path

### DISCUSSION THREAD

**Discussion by yerlaser** [nushell/nushell#5336] [Q&A] [answered]
I want to add paths written in a simple text file into my PATH. Here's my attempt.

```
$env.PATH
let nup = $'($env.HOME)/.config/nupaths.txt'
if ($nup | path exists) {
    open --raw $nup | lines | each {|it|
        let-env PATH = ($env.PATH | append $'$it')
    }
}
$env.PATH
```
I put it in a .nu script file and run it.
The path printed at the end is the same as the path in the beginning, so nothing gets appended.
What am I doing wrong?

**Comment by kubouch** [comment, marked as answer]
That's because any environment changes are scoped. See https://www.nushell.sh/book/thinking_in_nushell.html#nushell-s-environment-is-scoped 

You need to use something like `let-env PATH = if ... { $env.PATH | append foo } else { $env.PATH }`. Instead of the `each` loop, you'd use `reduce` to collect the paths. You might also want to use `prepend` which puts the new paths to the beginning of the PATH instead of at the end.

**Comment by yerlaser** [comment]
For those who need the same here's my working solution:

```
let-env PATH = if ($'($env.HOME)/.config/nushell/nupaths.txt' | path exists) {
    $env.PATH | split row ':' | prepend (open --raw $'($env.HOME)/.config/nushell/nupaths.txt' | lines) | uniq
} else {
    $env.PATH
}
```

**Reply by kubouch** [reply]
A few tips for more cross-platform code: 
1. You can use `char esep` instead of `:` in `split row` since Windows uses `;`.
2. Instead of making a path like `$'foo/bar/($baz)'` you can do `[ foo bar $baz ] | path join` which will connect with `/` or `\` based on the OS.

**Reply by kubouch** [reply]
Also, `$env.PATH` should be a list by default, there shouldn't be a need to manually split it.

**Reply by yerlaser** [reply]
Thanks for the tips. My improved version is now this:
```
let nupaths = ([$env.HOME .config nushell nupaths.txt] | path join)
let-env PATH = if ($nupaths | path exists) {
    $env.PATH | split row (char esep) | prepend (open --raw $nupaths | lines) | uniq
} else {
    $env.PATH
}
```

**Reply by yerlaser** [reply]
> Also, `$env.PATH` should be a list by default, there shouldn't be a need to manually split it.

However, when I remove the split like you suggest, the path ends up being a mix of lines coming from text file and the original path in a single line separated by colons. Probably functionally it's the same, but looks weird.

**Reply by kubouch** [reply]
Oh, you're right, you're defining it in `env.nu`, right? When evaluating the file, the `$env.PATH` is still a string, that's true. We automatically convert it afterwards. So the split is correct in your case.
